### ISSUE TITLE
`obfuscated_if_else` doesn't check for side effects in subexpressions

### ISSUE THREAD

**Issue by samueltardieu** [rust-lang/rust-clippy#14034]
### Summary

The lint suggests changing `b.then_some(x).unwrap_or(y)` into `if b { x } else { y }` without checking that the evaluation of `x` and `y` (whose evaluation is short-circuited in the suggested change) doesn't have any side effect. 

### Reproducer

I tried this code:

```rust
#![allow(clippy::all)]
#![warn(clippy::obfuscated_if_else)]

fn allocate(counter: &mut u32) -> u32 {
    *counter += 1;
    *counter
}

fn main() {
    let mut counter = 0;
    let a = 3;
    let c = (a == 1)
        .then_some(allocate(&mut counter))
        .unwrap_or(allocate(&mut counter));
    assert_eq!(c, 2);
    assert_eq!(counter, 2);
}
```

which runs without error.

Clippy proposed to transform it into:

```rust
#![allow(clippy::all)]
#![warn(clippy::obfuscated_if_else)]

fn allocate(counter: &mut u32) -> u32 {
    *counter += 1;
    *counter
}

fn main() {
    let mut counter = 0;
    let a = 3;
    let c = if (a == 1) { allocate(&mut counter) } else { allocate(&mut counter) };
    assert_eq!(c, 2);
    assert_eq!(counter, 2);
}
```

which fails the test `assert_eq!(c, 2)` because only one call to `allocate` now took place.

At minimum, the applicability should be downgraded from `MachineApplicable` to `MaybeIncorrect`.

### Version

```text
rustc 1.86.0-nightly (049355708 2025-01-18)
binary: rustc
commit-hash: 049355708383ab1b9a1046559b9d4230bdb3a5bc
commit-date: 2025-01-18
host: x86_64-unknown-linux-gnu
release: 1.86.0-nightly
LLVM version: 19.1.7
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by samueltardieu** [comment]
(pinging @lapla-cogito since we talked about this)

**Comment by lapla-cogito** [comment]
note: This problem was discovered in https://github.com/rust-lang/rust-clippy/pull/14021. My suggestion is to change the applicability for the time being in this PR.
